### DISCUSSION TITLE
Unable to Kill WatchImmediate

### DISCUSSION THREAD

**Discussion by aidanwardman** [tauri-apps/tauri#11087] [Q&A]
Edit: Using Tauri v2

I am attempting to add dynamic watchers to a folder I select, and can then subsequently change.

I create the watcher given a particular path and callback function I want to use:

```
let singletonWatcher:UnwatchFn = ()=>{};

async function addWatcher(path:string, callBack: Function, type:string){
     // Store UnwatchFn in variable for later use and killing??
     singletonWatcher = await watchImmediate(path,(event) => {
          console.log("Watcher Event Triggered: "+JSON.stringify(event));
          callBack();
     });
}
```

Then I change folder and attempt to kill it prior to re-creating it:

```
async function removeWatcher(path:string, type:string){
    // Call the stored function to kill the watcher??
    singletonWatcher();
    // Remove the function from the variable??
    singletonWatcher = ()=>{};
}
```

However I just can't appear to ever stop watching the folder, and each call just adds another and another and another watcher, all of which get triggered when ever an event happens.

Given the documented example below, how do I actually call the "stopWatching" to actually kill the watcher?

```
import { watch, watchImmediate } from 'tauri-plugin-fs-watch-api'

// can also watch an array of paths
const stopWatching = await watch('/path/to/something', { recursive: true }, event => {
  const { type, payload } = event
})

const stopRawWatcher = await watchImmediate(['/path/a', '/path/b'], {}, event => {
  const { path, operation, cookie } = event
})
```

Forgive my ignorance or dumbness.

**Comment by FabianLars** [comment]
Hmm, what you showed here seems correct, simply calling stopRawWatcher is supposed to unregister/stop everything. Are there any errors in the devtools console maybe?

**Reply by aidanwardman** [reply]
If I hard code Start and Stop to a button each, they appear to function as intended. However if I try and Stop the watcher, then overwrite the variable with a newly declared watcher, it's almost as if the variable retains both watchers simultaneously and every time I add a subsequent new watcher it just adds to the pile.

So after 3,4,5 folder changes, I end up with 3,4,5 watchers triggering every time a file changes, not ideal haha.

**Reply by aidanwardman** [reply]
![image](https://github.com/user-attachments/assets/39e6f849-f16a-409b-ad57-2d8fa7faa491)
You can see here , from changing folders 6 times, then adding a file to the folder, it has triggered 6 watcher events, even though they should all be stopping and recreating the watcher on a reused variable (perhaps this is not ideal? but I do not know a better way to make dynamic watchers)
